lib/spack/spack/cmd/release_jobs.py

```diff
@@ -0,0 +1,486 @@
+        os_names = [current_system]
+    else:
+        os_names = [name for name in containers]
+
+    container_specs = {name: {'image': None, 'specs': []} for name in os_names}
+
+    # Collect together all the specs that should be concretized in each
+    # container so they can all be done at once, avoiding the need to
+    # run the docker container for each spec separately.
+    for spec in spec_set:
+        for osname in os_names:
+            container_info = containers[osname]
+            image = None if current_system else container_info['image']
+            if image:
+                container_specs[osname]['image'] = image
+            if 'compilers' in container_info:
+                found_at_least_one = False
+                for item in container_info['compilers']:
+                    container_compiler_spec = CompilerSpec(item['name'])
+                    if spec.compiler == container_compiler_spec:
+                        container_specs[osname]['specs'].append(spec)
+                        found_at_least_one = True
+                if not found_at_least_one:
+                    tty.warn('No compiler in {0} satisfied {1}'.format(
+                        osname, spec.compiler))
+
+    for osname in container_specs:
+        if container_specs[osname]['specs']:
+            image = container_specs[osname]['image']
+            specs = container_specs[osname]['specs']
+            get_spec_dependencies(specs, deps, spec_labels, image)
+
+    dependencies = deps
+    unstaged = set(spec_labels.keys())
+    stages = []
+
+    while deps:
+        depends_on = set(deps.keys())
+        next_stage = unstaged.difference(depends_on)
+        stages.append(next_stage)
+        unstaged.difference_update(next_stage)
+        deps = remove_satisfied_deps(deps, next_stage)
+
+    if unstaged:
+        stages.append(unstaged.copy())
+
+    return spec_labels, dependencies, stages
+
+
+def print_staging_summary(spec_labels, dependencies, stages):
+    if not stages:
+        return
+
+    tty.msg('Staging summary:')
+    stage_index = 0
+    for stage in stages:
+        tty.msg('  stage {0} ({1} jobs):'.format(stage_index, len(stage)))
+
+        for job in sorted(stage):
+            s = spec_labels[job]['spec']
+            tty.msg('    {0} -> {1}'.format(job, get_spec_string(s)))
+
+        stage_index += 1
+
+
+def compute_spec_deps(spec_list, stream_like=None):
+    deptype = all_deptypes
+    spec_labels = {}
+
+    specs = []
+    dependencies = []
+
+    def key_label(s):
+        return s.dag_hash(), "%s/%s" % (s.name, s.dag_hash(7))
+
+    def append_dep(s, d):
+        dependencies.append({
+            'spec': s,
+            'depends': d,
+        })
+
+    for spec in spec_list:
+        spec.concretize()
+
+        root_spec = get_spec_string(spec)
+
+        rkey, rlabel = key_label(spec)
+
+        for s in spec.traverse(deptype=deptype):
+            if not s.concrete:
+                s.concretize()
```

Comment:
I'm not sure I follow.  Is it true that `s.traverse()` is generating the dependency specs of `spec`?  Until I have my hands on one of those dependency specs, how I can check if it's concrete?  I experimented with this code a bit to try and figure out how it worked, but I basically stole it from [here](https://github.com/spack/spack/blob/ec78d362d21d420fe6ee24bf888cb6fc2af37611/lib/spack/spack/graph.py#L562).  I added the concretize() in there myself though, because I wasn't getting all the dependencies without it.  Maybe we can chat about this one at some point if you have time.

Reply:
If X->Y->Z, and you concretize X with `X.concretize()`, all dependencies of X will also be concrete (that's part of the API of `X.concretize()` - we should update the docstring if that is unclear). If you think dependencies are missing from X after calling `X.concretize()` that is a bug in Spack.

Reply:
Ok, thanks for the clarification.  I'll find out for sure what's going on there, and file an issue if needed.

Reply:
I did some testing here, and it seems there was no problem removing the `concretize()` from within the `traverse()`, as you suggested.  So that's gone now, thanks.

Reply:
I still see that logic here, has it not been pushed yet?

Reply:
Ah for pete's sake.  Maybe a `git` mistake on my part?  I'll fix it, sorry about that.